src/KubernetesClient/Informers/Cache/VersionPartitionedSharedCache.cs

```diff
@@ -0,0 +1,294 @@
+using System;
+using System.Collections;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace k8s.Informers.Cache
+{
+    /// <summary>
+    /// Allows creating cache partitions for objects that have versioning semantics. Each partition will maintain its own view of its tracked objects,
+    /// but any items with same key and version will be shared across multiple cache partitions.
+    /// </summary>
+    /// <remarks>
+    /// The semantics of this class allows for object reuse between informers without compromising each informers ownership of its own cache. Primarily the issue
+    /// it solves is if multiple informers are created with different options, but the data they receive may overlap
+    /// (ex. overlapping labels, or informer scoped to namespace and another scoped globally). Since the master informer (actual connection to physical server) will receive
+    /// same notification over separate channels, we run the risk of informer desynchronization if they share the same cache. However, if each informer maintains it's own
+    /// cache, we may get multiple duplicate objects in memory. This allows any objects that share the same key/version to point to the same reference, while maintaining
+    /// integrity of each cache (dictionary). Note that unlike a regular dictionary, this does not allow updates to same key/version
+    /// </remarks>
+    /// <typeparam name="TKey">The type of the key uniquely identifying object</typeparam>
+    /// <typeparam name="TResource">The type of resource</typeparam>
+    /// <typeparam name="TVersion">The type of version associated with object</typeparam>
+    public class VersionPartitionedSharedCache<TKey, TResource, TVersion>
+    {
+        private readonly Func<TResource, TKey> _keySelector;
+        private readonly Func<TResource, TVersion> _versionSelector;
+        // internal to allow for unit testing
+        internal readonly Dictionary<VersionResourceKey,TResource> Items = new Dictionary<VersionResourceKey, TResource>();
+        private readonly HashSet<CacheView> _views = new HashSet<CacheView>();
+        private readonly object _lock = new object();
```

Comment:
All the lock semantics in this class are tricky, would it be better to use a concurrent collection?

Reply:
The reason locks are used and whole main reason for icache is to support atomic Reset operation.